DNSmonster ASIM DNS Parser

fhp-kkuphal opened the pull request:
Required items, please complete
   
   Change(s):
   - Added ASIM DNS Parser for DNSmonster

   Reason for Change(s):
   - New functionality

   Testing Completed:
   - Yes

   Checked that the validations are passing and have addressed any issues that are present:
   - Yes

fhp-kkuphal commented:
As this is my first contribution, I'm unclear if I should execute the KqlFuncYaml2Arm.py in my repository and include the resulting ARM files and changes to FullDeploymentDns.json as part of this pull request or if that is done elsewhere after the merge.

oshezaf commented on Parsers/ASimDns/Parsers/ASimDnsDnsMonster.yaml:
```diff
@@ -0,0 +1,52 @@
+    dnsmonster_CL | where not(disabled)
+    | where DNS_Response_b
+    | extend DNS_Question = todynamic(DNS_Question_s)
+    | mv-expand DNS_Question
```
Using mv-expand means that you will get multiple events, one for each DNS question. 

The DNS schema does not support multiple questions in a single event because after researching it we found that while a valid DNS optoin, it is never really used. Did you encounter a real-world example of multiple questions in a single DNS query?

fhp-kkuphal replied:
No.  I may have misunderstood the function then.  DNS_Question contains values such as

```
"[
  {
    ""Name"": ""p13n.adobe.io."",
    ""Qtype"": 1,
    ""Qclass"": 1
  }
]"
```

Which I need to use later as DNS_Question.Qtype and DNS_Question.QClass.  mv-expand was a way to break out those values but I can just easily reference as DNS_Question[0].Qtype without the mv-expand.  As you indicated, I have not seen multiple Questions in a single query.

oshezaf commented on Parsers/ASimDns/Parsers/vimDnsDnsMonster.yaml:
```diff
@@ -0,0 +1,97 @@
+    | where (isnull(starttime) or TimeGenerated >= starttime)
+        and (isnull(endtime) or TimeGenerated <= endtime)
+        and (srcipaddr=='*' or SrcIP_s==srcipaddr)
+        and (response_has_ipv4=='*' or ipv4_is_in_range(tostring(DNS_Answer.A),response_has_ipv4))
```
DNS_Answer should be assigned to DnsReponseName

oshezaf commented on Parsers/ASimDns/Parsers/vimDnsDnsMonster.yaml:
```diff
@@ -0,0 +1,97 @@
+    | mv-expand DNS_Question
+    | mv-expand DNS_Answer
+    | extend EventType = opcodes[toint(DNS_Opcode_d)],
+        EventSubType = "request",
```
It is most probably response, especially if answer is available.

oshezaf commented on Parsers/ASimDns/Parsers/vimDnsDnsMonster.yaml:
```diff
@@ -0,0 +1,97 @@
+    | where DNS_Response_b
+    | extend DNS_Question = todynamic(DNS_Question_s), DNS_Answer = todynamic(DNS_Answer_s)
+    | mv-expand DNS_Question
+    | mv-expand DNS_Answer
```
Since the event includes a list of questions (though I suspect only one) and a list of answers, this dual mv-expand will create a cartesian multiplication which is not what you wanted. While for question, our analysis is that it is always one, for the answer, we just handle it as a single string in DnsResponseName.

fhp-kkuphal replied:
As noted above, I can remove the mv-expand for DNS_Question easily enough.  The DNS_Answer does routinely contain multiple values.  Is there a problem with the DNS_Answer expansion in order to perform the has_ipv4_prefix and ipv4_is_in_range tests against each result?

oshezaf replied:
Our solution to balance performance and usability was to leave the answer field unparsed and to use the ipv4 functions on the unparsed field. They work well.  

Few side notes:
- If you want to handle a JSON array, but not duplicate events, look into mv-apply.
- That said, JSON arrays are hard to use in a query. As a parser provider, we avoid creating them.

oshezaf commented on Parsers/ASimDns/Parsers/vimDnsDnsMonster.yaml:
```diff
@@ -0,0 +1,97 @@
+    | mv-expand DNS_Answer
+    | extend EventType = opcodes[toint(DNS_Opcode_d)],
+        EventSubType = "request",
+        EventResultDetails = rcodes[toint(DNS_Opcode_d)],
```
You need to add aliases, for example DnsResponseCodeName for EventResultDetails

oshezaf reviewed (changes requested):
Hi Kevin (@fhp-kkuphal ). 

I left a few comments, but they are not comprehensive. 

To review in detail:
- Did you run the ASIM tests on the parser? Based on my review, there will be errors. A valid parser is one that has no errors, and document exceptions for warnings.
- Can you provide [sample data](https://github.com/Azure/Azure-Sentinel/tree/master/Sample%20Data/Custom) so I can do a better job of reviewing?

Also, yours is the first community contribution to ASIM. Until now, it was either Microsoft developers or a partner creating a parser for their product. So, I need to formulate our process for that.

To that end, the first question is:
- Do you want it to part of the ASIM package, so deployed by default to every customer. I would love it, but it requires more work on your side. No worries, I will tell you what's needed.
- Or, you prefer to keep it as a standalone parsers, that customers deploy separately and add to the ASIM union parsers themselves? It has the advantage of less requirements from my side. If so, we will need to have it in a different location in the repository. I need to come up with the right path

~ Ofer

fhp-kkuphal commented:
> * Did you run the ASIM tests on the parser? Based on my review, there will be errors. A valid parser is one that has no errors, and document exceptions for warnings.

I did not.  I have updated the PR and will conduct those tests.

> * Can you provide [sample data](https://github.com/Azure/Azure-Sentinel/tree/master/Sample%20Data/Custom) so I can do a better job of reviewing?

I added a set of sample data in aae59ee

> * Do you want it to part of the ASIM package, so deployed by default to every customer. I would love it, but it requires more work on your side. No worries, I will tell you what's needed.

I would be OK with that but really, I'm open to whatever works best.  I'm not sure how widespread the use of DnsMonster would be but since it implements [dnstap](https://dnstap.info/) in conjunction with Sentinel output, it solves a problem that we encountered by having [DNSdist](https://dnsdist.org/) load balancing in front of our Microsoft DNS which resulted in the masking of the true client IP.  Instead of loading our Microsoft DNS into Sentinel, we are tapping DNSdist through DnsMonster and getting workable DNS data in Sentinel that we were missing before.

v-spadarthi commented:
@fhp-kkuphal : We observed this PR is in Draft state since long time. Could you please move this PR from Draft to Open ?

v-spadarthi commented:
@fhp-kkuphal : We observed this PR is in Draft state since long time. Could you please move this PR from Draft to Open ?

v-spadarthi commented:
@oshezaf : Could you please review and provide your signoff.Thanks!!

oshezaf commented:
Hi @fhp-kkuphal, happy to get back to your contribution. One question that comes to mind is what is the connector you are using to get the events to Sentinel. Would it be [this one](https://dnsmonster.dev/docs/outputs/sentinel/)?

v-mchatla commented:
Hi @oshezaf,
Author has addressed your question. Can you please review and provide your sign off.
Thanks

v-mchatla commented:
Hi @oshezaf,
Sorry for multiple follow ups, can you please review and provide your approval.
Thanks

v-mchatla commented:
Hi @oshezaf,
Can you please review and provide your approval.
Thanks

oshezaf commented:
Hi @fhp-kkuphal  and @v-mchatla : the parser needs additional work. To not overburden with comments I have cloned @fhp-kkuphal's repo and will PR to that, so the you can merge and push up. It will take me some time to finish though.

devikamehra commented:
@fhp-kkuphal Wanted to check if you are still interested in making this contribution. If so, please follow the updated guidelines recently added here - https://github.com/Azure/Azure-Sentinel/blob/master/Parsers/Readme.md

v-atulyadav commented:
We wanted to check on the status of PR #4544. PR is pending for more than 30 days. Please let us know if you need any assistance to review this PR. Per our standard operating procedures if no response is received in the next 7 business days, we will close this PR. Thank you for your cooperation.

v-atulyadav commented:
Hi @fhp-kkuphal, since we have not received a response in the last 7 days, we are closing your PR #4544 per our standard operating procedures. If you still need support for this issue, you can re-open the PR at any time.
If you do re-open, we simply request that you ensure the PR has response to the last request. Thank you for your cooperation
